File: src/modules/graphs/3_cycleDetection_Undirected/BFS.ts
class Solution {
    detect(src: number, adj: number[][], vis: number[]): boolean {
        vis[src] = 1; // make it visited
        const queue: [number, number][] = [];
        queue.push([src, -1]); // storing [value,previous]

        while (queue.length > 0) {
            const [node, parent] = queue.shift()!;

            for (const neighbor of adj[node]) {
                if (!vis[neighbor]) {
                    vis[neighbor] = 1;
                    queue.push([neighbor, node]);
                } else if (parent !== neighbor) {
                    return true;
                }
            }
        }
        return false;
    }

    isCycle(adj: number[][]): boolean {
        const n = adj.length;
        const visited: number[] = new Array(n).fill(0); // visited matrix 

        for (let i = 0; i < n; i++) {
            if (!visited[i]) {
                if (this.detect(i, adj, visited)) return true; // if any one is the component found with a cycle, it has a cycle
            }
        }
        return false;
    }
}

export { }

File: src/modules/graphs/3_cycleDetection_Undirected/bfs.go
package main

type relation struct {
	node   int
	parent int
}

/*
@approach
1. We'll start a bfs traversal on the graph and we keep exploring
*/
func checkCycle(src int, adj [][]int, visited []bool) bool {
	visited[src] = true
	queue := []*relation{}
	queue = append(queue, &relation{node: src, parent: -1})

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]

		for _, neighbor := range adj[curr.node] {
			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, &relation{node: neighbor, parent: curr.node})
			} else if curr.parent != neighbor {
				// if it is already visited and it's not even the parent
				return true
			}
		}
	}
	return false
}

func isCycle(adj [][]int) bool {
	n := len(adj)
	visited := make([]bool, n)

	for i := 0; i < n; i++ {
		if !visited[i] {
			if checkCycle(i, adj, visited) {
				return true
			}
		}
	}
	return false
}
